graphql_compiler/compiler/ir_lowering_cypher/ir_lowering.py

```diff
@@ -89,16 +92,89 @@ def remove_mark_location_after_optional_backtrack(ir_blocks, query_metadata_tabl
     return new_ir_blocks
 
 
+class CypherFoldScopeContextField(Expression):
+    """An expression for a field captured in a @fold scope for non _x_count filtering."""
+
+    __slots__ = ('fold_scope_location', 'field_type')
+
+    def __init__(self, fold_scope_location, field_type):
+        """Construct a new CypherFoldScopeContextField object for this folded field.
+
+        Args:
+            fold_scope_location: FoldScopeLocation specifying the location of
+                                 the context field being output.
+            field_type: GraphQL type object, specifying the type of the field being output.
+
+        Returns:
+            new CypherFoldScopeContextField object
+        """
+        super(CypherFoldScopeContextField, self).__init__(fold_scope_location, field_type)
+        self.fold_scope_location = fold_scope_location
+        self.field_type = field_type
+        self.validate()
+
+    def validate(self):
+        """Validate that the CypherFoldScopeContextField is correctly representable."""
+        if not isinstance(self.fold_scope_location, FoldScopeLocation):
+            raise TypeError(u'Expected FoldScopeLocation fold_scope_location, got: {} {}'.format(
+                type(self.fold_scope_location), self.fold_scope_location))
+
+        if self.fold_scope_location.field is None:
+            raise ValueError(u'Expected FoldScopeLocation at a field, but got: {}'
+                             .format(self.fold_scope_location))
+
+        if self.fold_scope_location.field == COUNT_META_FIELD_NAME:
+            raise TypeError(u'Expected fold_scope_location field to not be the _x_count meta-field '
+                            u'because CypherFoldedContextField is specifically for filtering on '
+                            u'individual vertices\' fields in a fold scope, not for filtering on '
+                            u'the size of the list. Got FoldScopeLocation: {}'
+                            .format(self.fold_scope_location))
+
+        if not is_graphql_type(self.field_type):
+            raise ValueError(u'Invalid value of "field_type": {}'.format(self.field_type))
+
+    def to_gremlin(self):
+        """Should never be called, since this is Cypher-specific."""
+        raise NotImplementedError()
+
+    def to_match(self):
+        """Should never be called, since this is Cypher-specific."""
+        raise NotImplementedError()
+
+    def to_cypher(self):
+        """Return a unicode object with the Cypher representation of this expression."""
+        self.validate()
+
+        _, field_name = self.fold_scope_location.get_location_name()
+        mark_name = emit_cypher.fold_scope_location_full_path_name(self.fold_scope_location)
```

Comment:
The fact that we have to reach from the Cypher lowering code into the Cypher emit code seems to suggest that this helper function is in the wrong place. Let's put it in a file where we don't have to break abstraction boundaries to import it.